Custom root certificates for JWKSResolver

dgn opened the pull request:
[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[x] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Fixes #15086, should also work for #9505

This PR adds a variable to the `pilot` helm chart to inject a trusted CA into the JWKSResolver. This can be necessary when hosting your JWKS on a server protected with a certificate from an internal CA.

When installing via `helm`, you can now specify a PEM-encoded certificate in `pilot.jwksResolverExtraRootCA` in your `values.yaml` to trust an additional CA certificate:
```yaml
...
pilot:
  enabled: true
  jwksResolverExtraRootCA: |
    -----BEGIN CERTIFICATE-----
    [...]
    -----END CERTIFICATE-----
...
```

Or, during installation you can directly specify your certificate as a command-line argument:

```bash
helm install [...] --set-file pilot.jwksResolverExtraRootCA=yourRootCertificate.pem
```

dgn committed:
pilot: read additional CA bundles from /cacerts/extra.pem if present

dgn committed:
install: updated pilot helm chart to create ConfigMap if pilot.jwksResolverExtraRootCA is set

istio-testing commented:
Hi @dgn. Thanks for your PR.

I'm waiting for a [istio](https://github.com/orgs/istio/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should [join the org](https://github.com/orgs/istio/people) to skip this step.

Once the patch is verified, the new status will be reflected by the `ok-to-test` label.

I understand the commands that are listed [here](https://go.k8s.io/bot-commands?repo=istio%2Fistio).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

dgn committed:
pilot: add func to make NewJWKSResolver easily testable with custom certs

Also, make the strings that hold the paths to CA bundles const, remove
previously added error message

dgn committed:
pilot: add unit tests for resolving JWKS from HTTPS URIs

dgn committed:
pilot: add testing certificates for jwks unit tests

dgn committed:
install: add new configmap template

morvencao commented:
/ok-to-test

istio-testing commented:
@dgn: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-mixer-k8s-tests_istio | b5b69fef7ec259da5845dc1575adaadd09189c98 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/17176/integ-mixer-k8s-tests_istio/495) | `/test integ-mixer-k8s-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

dgn commented:
/retest

rshriram commented on install/kubernetes/helm/istio/charts/pilot/templates/deployment.yaml:
```diff
@@ -214,6 +218,11 @@ spec:
+{{- if .Values.jwksResolverExtraRootCA }}
+      - name: extracacerts
+        configMap:
+          name: istio-jwks-extra-cacerts
```
why is this extra? shouldnt it be istio-jwks-root-ca ?

dgn replied:
Well, we include a bundle of standard CA certificates in the container, anything you add here will *also* be trusted, in addition to that included bundle.

pitlv2109 commented:
@geeknoid @howardjohn This is tagged as `area/security` (indeed it is), can we do something so that bot will request code review from the security group?

howardjohn commented:
> This is tagged as area/security (indeed it is), can we do something so that bot will request code review from the security group?

That would be nice. I suspect its pretty infrequent to have a security PR that doesn't require security owners approval though

howardjohn commented on pilot/pkg/model/test/testcert/generate.sh:
```diff
@@ -0,0 +1,23 @@
+        -days 3650 -nodes -out cert.pem -keyout key.pem \
+        -subj "/C=US/ST=Denial/L=Ether/O=Dis/CN=localhost"
+
+openssl req -new -newkey rsa:4096 -x509 -sha256 \
```
thank you for adding this

howardjohn reviewed (commented):
lgtm, will wait for security folks to approve.

@yangminzhu

yangminzhu reviewed (approved):
Thanks for the PR.
Could you add instructions of how to use the extra CA bundle in the PR description and also make sure to update the 1.4 release notes for this?

yangminzhu commented:
@howardjohn Could you approve if this looks good to you? Thanks!

howardjohn reviewed (approved):
Please send a PR to istio/install so this can be used from the operator

howardjohn commented:
/test gencheck_istio

istio-testing commented:
@dgn: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-security-k8s-tests_istio | b5b69fef7ec259da5845dc1575adaadd09189c98 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/17176/integ-security-k8s-tests_istio/2384) | `/test integ-security-k8s-tests_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

howardjohn commented:
/retest

istio-testing commented:
@dgn: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-security-k8s-tests_istio | b5b69fef7ec259da5845dc1575adaadd09189c98 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/17176/integ-security-k8s-tests_istio/2384) | `/test integ-security-k8s-tests_istio`
integ-telemetry-k8s-tests_istio | b5b69fef7ec259da5845dc1575adaadd09189c98 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/17176/integ-telemetry-k8s-tests_istio/2297) | `/test integ-telemetry-k8s-tests_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

dgn commented:
/retest
